src/rust/src/x509.rs

```diff
@@ -544,6 +547,441 @@ fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<
+            .getattr("Encoding")?;
+
+        let result = asn1::write_single(self.raw.borrow_value());
+        if encoding == encoding_class.getattr("DER")? {
+            Ok(pyo3::types::PyBytes::new(py, &result))
+        } else if encoding == encoding_class.getattr("PEM")? {
+            let pem = pem::encode_config(
+                &pem::Pem {
+                    tag: "CERTIFICATE REQUEST".to_string(),
+                    contents: result,
+                },
+                pem::EncodeConfig {
+                    line_ending: pem::LineEnding::LF,
+                },
+            )
+            .into_bytes();
+            Ok(pyo3::types::PyBytes::new(py, &pem))
+        } else {
+            Err(pyo3::exceptions::PyTypeError::new_err(
+                "encoding must be Encoding.DER or Encoding.PEM",
+            ))
+        }
+    }
+
+    fn get_attribute_for_oid<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        oid: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        // TODO: more efficient way to do this? Possibly convert ObjectIdentifier in python to a rust class.
+        let oid_str = oid
+            .getattr("dotted_string")?
+            .downcast::<pyo3::types::PyString>()?
+            .to_string();
+        let rust_oid = asn1::ObjectIdentifier::from_string(&oid_str).unwrap();
+        // TODO: no more unwrap once attributes isn't an Optional
+        for attribute in self.raw.borrow_value().csr_info.attributes.clone().unwrap() {
+            if rust_oid == attribute.type_id {
+                check_attribute_length(attribute.values.clone())?;
+                if let Some(val) = attribute.values.clone().next() {
+                    // We allow utf8string, printablestring, and ia5string at this time
+                    if val.tag() == 0x0c || val.tag() == 0x13 || val.tag() == 0x16 {
+                        return Ok(pyo3::types::PyBytes::new(py, val.data()));
+                    } else {
+                        return Err(pyo3::exceptions::PyValueError::new_err(format!(
+                            "OID {} has a disallowed ASN.1 type: {}",
+                            oid,
+                            val.tag()
+                        )));
+                    }
+                }
+            }
+        }
+        Err(pyo3::PyErr::from_instance(
+            py.import("cryptography.x509")?.call_method1(
+                "AttributeNotFound",
+                (
+                    format!("No {} attribute was found", rust_oid.to_string()),
+                    oid,
+                ),
+            )?,
+        ))
+    }
+
+    #[getter]
+    fn extensions(&mut self, py: pyo3::Python<'_>) -> pyo3::PyResult<pyo3::PyObject> {
+        let x509_module = py.import("cryptography.x509")?;
+        let exts = self.raw.borrow_value().csr_info.get_extension_attribute()?;
+        parse_and_cache_extensions(py, &mut self.cached_extensions, &exts, |oid, ext_data| {
```

Comment:
This duplicates a ton of cert parsing, can we put it in a function that we reuse, as we do for OCSP/CRLEntry

Reply:
(PS: Put that refactoring in it's own PR for ease of review plz)

Reply:
This is an overlapping set but there are two more extensions in cert via csr so we'll have to handle that small variance